torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -212,13 +212,14 @@ def rng_functionalization_wrapper(args: List[Any]):
     if not hasattr(compiled_fw_func, "_boxed_call"):
         compiled_fw_func = make_boxed_func(compiled_fw_func)
 
-    compiled_fn = create_runtime_wrapper(
-        compiled_fw_func,
-        runtime_metadata=fw_metadata,
+    compiled_fn = RuntimeWrapper(
         indices_of_inps_to_detach=[],
         trace_joint=False,
-        keep_input_mutations=aot_config.keep_inference_input_mutations,
         disable_amp=disable_amp,
+    ).post_compile(
```

Comment:
Mostly just a callout: for the most part we have a 1:1 relationship between "`pre_compile` and `post_compile` transformations. But for the particular case of this `runtime_wrapper` bit, the transformation is something like:

in the inference path: `aot_dispatch_base()` makes the compile-time change, and uses `create_runtime_wrapper` to make the runtime change

in the training path: `aot_dispatch_autograd()` makes the compile-time change, and uses a combination of `CompiledFunction` + `create_runtime_wrapper` to make the runtime change.

So the actual "post_compile" runtime wrapper change needed between the inference and training paths have some commonalities, that are shared in the the `create_runtime_wrapper` helper.

Reply:
If we are refactoring all of the runtime wrappers to use `CompilerWrapper`, then I'm mostly just calling out that I would have originally expected every instance to have both a `pre_compile` and `post_compile`, and I guess we have just a `post_compile` here since it lines up with how the runtime wrappers are used today / makes the refactor easier.

Do you imagine a better end-state being that we eventually refactor the wrapper logic so there is more of a 1:1 between pre-compile and post-compile for each layer? Or do you think this is ok as an end state (I think this refactor looks good to me though, happy to stamp)

Reply:
Hmm, so from what I can tell, the runtime change is only made in create_runtime_wrapper, is it not? Do you mind giving an example of somewhere where CompiledFunction is making a runtime change related to what create_runtime_wrapper is doing? I had thought that the only runtime change create_runtime_wrapper is responsible for was in the two places it's called in aot_dispatch_autograd and aot_dispatch_base, both of which are refactored here. 

Just based on my own understanding, I had thought that create_runtime_wrapper was a runtime only wrapper (in that it does not seem to affect the flat_fn state), but if you could point me to what you would consider the "pre_compile" step of create_runtime_wrapper (i.e., where it's modifying the flat_args or flat_fn), I can follow up and refactor that bit too!

Reply:
Thinking about this more, I think the point you're makign is that aot_dispatch_base and aot_dispatch_autograd, are, in a sense, themselves just big wrappers, and `create_runtime_wrapper` is actually just a post_compile *for* the precompile step that exists somewhere in the code of aot_dispatch_*. 

I do think it would be valuable to disentangle the logic of aot_dispatch_* that has to do with whatever create_runtime_wrappers is doing so that it can be expressed as a pre_compile for the new wrapper. Happy to work on that, though I might prioritize it a bit lower just because it's not necessary for caching.

I think the same can be said about the other "post compile only" wrappers that I define in the next PR: there's probably some pre-compile logic in aot_dispatch_* that could be put into a pre compile step, but it would require a more involved refactor. I even tried to do it for rng functionalization, but found it to be pretty challenging to get right just due to the sheer number of tangled dependencies within aot_dispatch_autograd.

Reply:
> but found it to be pretty challenging to get right just due to the sheer number of tangled dependencies within aot_dispatch_autograd.

yeah... agreed 😛